quickjs/src/jniMain/kotlin/app/cash/quickjs/ZiplineJvm.kt

```diff
@@ -0,0 +1,101 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.quickjs
+
+import app.cash.quickjs.internal.bridge.InboundService
+import app.cash.quickjs.internal.bridge.InternalBridge
+import app.cash.quickjs.internal.bridge.KtBridge
+import app.cash.quickjs.internal.bridge.OutboundClientFactory
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.delay
+
+internal class ZiplineJvm(
+  override val quickJs: QuickJs,
+  override val dispatcher: CoroutineDispatcher,
+) : Zipline(), HostPlatform {
+  private val ktBridge: KtBridge
+  private val jsPlatform: JsPlatform
+
+  init {
+    val jsInboundBridge = object : InternalBridge {
+      // Lazily fetch the bridge to call them.
+      val delegate: InternalBridge by lazy(mode = LazyThreadSafetyMode.NONE) {
+        quickJs.get(
+          name = "app_cash_quickjs_inboundBridge",
+          type = InternalBridge::class.java
+        )
+      }
+
+      override fun invoke(
+        instanceName: String,
+        funName: String,
+        encodedArguments: ByteArray
+      ): ByteArray {
+        return delegate.invoke(instanceName, funName, encodedArguments)
+      }
+
+      override fun invokeSuspending(
+        instanceName: String,
+        funName: String,
+        encodedArguments: ByteArray,
+        callbackName: String
+      ) {
+        return delegate.invokeSuspending(instanceName, funName, encodedArguments, callbackName)
+      }
+    }
+
+    ktBridge = KtBridge(dispatcher, outboundBridge = jsInboundBridge)
+
+    // Eagerly publish the bridge so they can call us.
+    quickJs.set(
+      name = "app_cash_quickjs_outboundBridge",
+      type = InternalBridge::class.java,
+      instance = ktBridge.inboundBridge
+    )
+
+    // Connect platforms using our newly-bootstrapped bridges.
+    ktBridge.set<HostPlatform>(
+      name = "app.cash.quickjs.hostPlatform",
+      jsAdapter = BuiltInJsAdapter,
+      instance = this
+    )
+    jsPlatform = ktBridge.get(
+      name = "app.cash.quickjs.jsPlatform",
+      jsAdapter = BuiltInJsAdapter
+    )
+  }
+
+  override val engineVersion: String
+    get() = quickJsVersion
+
+  override fun <T : Any> get(
+    name: String,
+    outboundClientFactory: OutboundClientFactory<T>
+  ): T {
+    return ktBridge.get(name, outboundClientFactory)
+  }
+
+  override fun set(name: String, handler: InboundService<*>) {
+    ktBridge.set(name, handler)
+  }
+
+  override fun setTimeout(timeoutId: Int, delayMillis: Int) {
+    launch {
```

Comment:
I'm wondering if this should have `start=UNDISPATCHED` so that we get to the delay before the function returns. But I'm not sure what the expected behavior is around a 0 delayMillis for `setTimeout` and what the `delay(0)` actual behavior is.

Reply:
I think you’re on to something. Lemme do, but in follow-up with new tests.